File: backend/utils/currency.py
"""
Currency conversion & formatting for FinSight.
Fetches live exchange rates from open.er-api.com with 1-hour cache.
"""

import json
import logging
import urllib.request
from datetime import datetime, timedelta

logger = logging.getLogger("finsight.currency")

RATES = {}
LAST_FETCH = None
CACHE_DURATION = timedelta(hours=1)

HARDCODED_RATES = {
    "INR": 1.0,
    "USD": 0.012,
    "NPR": 1.60,
}

SYMBOLS = {
    "INR": "\u20B9",
    "USD": "$",
    "NPR": "\u0930\u0941",
}

PLAIN = {
    "INR": "Rs.",
    "USD": "$",
    "NPR": "Rs.",
}

CURRENCY_NAMES = ["INR", "USD", "NPR"]


def get_rates() -> dict:
    global RATES, LAST_FETCH
    now = datetime.utcnow()
    if LAST_FETCH and now - LAST_FETCH < CACHE_DURATION and RATES:
        return RATES
    try:
        req = urllib.request.Request(
            "https://open.er-api.com/v6/latest/INR",
            headers={"User-Agent": "FinSight/2.0"},
        )
        with urllib.request.urlopen(req, timeout=10) as resp:
            data = json.loads(resp.read().decode())
        rates = {"INR": 1.0}
        for cur in CURRENCY_NAMES:
            if cur != "INR" and cur in data.get("rates", {}):
                rates[cur] = data["rates"][cur]
        RATES = rates
        LAST_FETCH = now
        logger.info("Exchange rates updated from live API")
    except Exception:
        logger.warning("Failed to fetch live rates, using hardcoded fallback")
        RATES = HARDCODED_RATES.copy()
        LAST_FETCH = now
    return RATES


def _to_float(val) -> float:
    return float(val)


def convert_to_inr(amount, from_currency: str) -> float:
    amount = _to_float(amount)
    if not from_currency or from_currency == "INR":
        return round(amount, 2)
    rates = get_rates()
    rate_to_inr = 1.0 / rates.get(from_currency, 1.0)
    return round(amount * rate_to_inr, 2)


def convert(amount_inr, to_currency: str) -> float:
    amount_inr = _to_float(amount_inr)
    rates = get_rates()
    rate = rates.get(to_currency, 1.0)
    return round(amount_inr * rate, 2)


def format_amount(amount_inr, currency: str, plain: bool = False) -> str:
    converted = convert(amount_inr, currency)
    sym = PLAIN.get(currency, currency) if plain else SYMBOLS.get(currency, currency)
    return f"{sym}{converted:,.2f}"


def get_conversion_note(from_cur: str, to_cur: str) -> str:
    one_converted = convert(1.0, to_cur)
    return f"1 {from_cur} \u2248 {one_converted:.4f} {to_cur}"

File: server/utils/currency.js
let ratesCache = {};
let lastFetch = null;
const CACHE_DURATION = 60 * 60 * 1000; // 1 hour in ms

const HARDCODED_RATES = {
  INR: 1.0,
  USD: 0.012,
  NPR: 1.60
};

const SYMBOLS = {
  INR: '₹',
  USD: '$',
  NPR: 'रु'
};

const CURRENCY_NAMES = ['INR', 'USD', 'NPR'];

async function getRates() {
  const now = Date.now();
  if (lastFetch && (now - lastFetch < CACHE_DURATION) && Object.keys(ratesCache).length > 0) {
    return ratesCache;
  }

  try {
    const response = await fetch('https://open.er-api.com/v6/latest/INR');
    if (!response.ok) {
      throw new Error(`API returned status ${response.status}`);
    }
    const data = await response.json();
    
    const rates = { INR: 1.0 };
    for (const cur of CURRENCY_NAMES) {
      if (cur !== 'INR' && data.rates && data.rates[cur]) {
        rates[cur] = data.rates[cur];
      }
    }
    ratesCache = rates;
    lastFetch = now;
  } catch (error) {
    console.warn('Failed to fetch live rates, using hardcoded fallback:', error.message);
    ratesCache = { ...HARDCODED_RATES };
    lastFetch = now;
  }

  return ratesCache;
}

function getConversionNote(fromCur, toCur, rates) {
  const toRate = rates[toCur] || 1.0;
  const fromRate = rates[fromCur] || 1.0;
  const rate = toRate / fromRate;
  return `1 ${fromCur} ≈ ${rate.toFixed(4)} ${toCur}`;
}

module.exports = {
  getRates,
  getConversionNote,
  SYMBOLS,
  CURRENCY_NAMES
};
